plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java

```diff
@@ -142,27 +150,33 @@
 	@Parameter(property = LicenseHeaderStep.spotlessSetLicenseHeaderYearsFromGitHistory)
 	private String setLicenseHeaderYearsFromGitHistory;
 
-	protected abstract void process(Iterable<File> files, Formatter formatter) throws MojoExecutionException;
+	// todo: check naming convensions for plugin properties
+	@Parameter(property = "spotless.upToDateCheckingEnabled", defaultValue = "false")
+	private boolean upToDateCheckingEnabled;
+
+	protected abstract void process(Iterable<File> files, Formatter formatter, UpToDateChecker upToDateChecker) throws MojoExecutionException;
 
 	@Override
 	public final void execute() throws MojoExecutionException {
-		List<FormatterFactory> formatterFactories = getFormatterFactories();
-		for (FormatterFactory formatterFactory : formatterFactories) {
-			execute(formatterFactory);
-		}
-	}
-
-	private void execute(FormatterFactory formatterFactory) throws MojoExecutionException {
 		if (skip) {
 			getLog().info(String.format("Spotless %s skipped", goal));
 			return;
 		}
 
+		List<FormatterFactory> formatterFactories = getFormatterFactories();
 		FormatterConfig config = getFormatterConfig();
-		List<File> files = collectFiles(formatterFactory, config);
 
-		try (Formatter formatter = formatterFactory.newFormatter(files, config)) {
-			process(files, formatter);
+		Map<FormatterFactory, List<File>> formatterFactoryToFiles = new HashMap<>();
```

Comment:
Can this be `Supplier<Iterator<File>>`, or at least `Supplier<List<File>>`? The point of this PR is to improve performance on very large builds, stuffing every path into memory at exactly the same time isn't ideal.

Reply:
Edit: I think I'm wrong about this because it's actually part of how the maven plugin handles `GitRatchet`, see [this followup comment](https://github.com/diffplug/spotless/pull/935/files#r744349033).